src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,50 +1,167 @@
+        private Node<T> prev;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.value = value;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private T unlink(Node<T> removeNode) {
+        Node<T> nextNode = removeNode.next;
+        Node<T> prevNode = removeNode.prev;
+        if (prevNode == null) {
+            head = nextNode;
+        } else {
+            prevNode.next = nextNode;
+            removeNode.prev = null;
+        }
+
+        if (nextNode == null) {
+            tail = prevNode;
+        } else {
+            nextNode.prev = prevNode;
+            removeNode.next = null;
+        }
+        T valueNode = removeNode.value;
+        removeNode.value = null;
+        size--;
+        return valueNode;
+    }
+
+    private boolean isAvailableIndex(int index) {
+        return index >= 0 && index < size;
+    }
+
+    private void linkLast(T value) {
+        Node<T> lastNode = tail;
+        Node<T> newNode = new Node(lastNode, value, null);
+        tail = newNode;
+        if (lastNode == null) {
+            head = newNode;
+        } else {
+            lastNode.next = newNode;
+        }
+        size++;
+    }
+
+    private Node<T> node(int index) {
+        Node<T> current;
+        if (index < size / 2) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+            return current;
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+            return current;
+        }
+    }
+
+    private void linkBefore(T value, Node<T> findIndexNode) {
+        Node<T> previous = findIndexNode.prev;
+        Node<T> newNode = new Node(previous, value, findIndexNode);
+        findIndexNode.prev = newNode;
+        if (previous == null) {
+            head = newNode;
+        } else {
+            previous.next = newNode;
+        }
+        size++;
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        linkLast(value);
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
+        if (!(index >= 0 && index <= size)) {
+            throw new IndexOutOfBoundsException("Index " + index + " is out of bound");
+        }
 
+        if (index == size) {
+            linkLast(value);
+        } else {
+            linkBefore(value, node(index));
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (int i = 0; i < list.size(); i++) {
+            T value = list.get(i);
+            add(value);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (!(isAvailableIndex(index))) {
+            throw new IndexOutOfBoundsException("Index " + index + " is out of bound");
```

Comment:
this check is repeated several times in the code, let's move it to a separate method

Reply:
done